HPCC-27122 Create Test Auth secmgr for ECLWatch

wangkx opened the pull request:
1. Add security plugin: testauthSecurity.
2. Use the security plugin as default auth for ECLWatch.
3. Read the setting for the security plugin.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-27122 Add Test Auth Service to ECLWatch

1. Add security plugin: testauthSecurity.
2. Use the security plugin as default auth for ECLWatch.
3. Read the setting for the security plugin.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

afishbeck commented on esp/platform/application_config.cpp:
```diff
@@ -151,6 +151,29 @@ bool addAuthNZSecurity(const char *name, IPropertyTree *legacyEsp, IPropertyTree
     return true;
 }
 
+bool addTestAuthSecurity(IPropertyTree *legacyEsp, IPropertyTree *appEsp, StringBuffer &bindAuth)
```
@wangkx now that I look at it, you shouldn't need "testauth" specific routines in application_config.

Since the config is based on the SecurityManagers/SecurityManager format you should be able to make use of the "authNZ" mechanism.

wangkx replied:
@afishbeck The "authNZ" mechanism works now.

afishbeck commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,544 @@
+    SecAccessFlags defaulteECLWUScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFileScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFeaturAccess = SecAccess_Unavailable;
+    MapStringTo<int> featureAccesses, eclwuScopeAccesses, fileScopeAccesses;
```
I think for fileScopes I would build a tree from all the scopes configured and save the access flag in the tree as well.  Then when I needed to authorize a scope I would parse the nodes and walk the tree until I had to authorize a node that wasn't in that part of the tree.  The last access flag found wins.  If no access flag was found  use default.

For config:
```
  defaults:
    filescope: NONE
  filescopes:
    "a": FULL
    "a::b::c": READ
    "a::x": READ
    "z::d": FULL
    "z::d::e": NONE
```
Load into a PTREE instead of a MapStringTo:
```
<root access=NONE> <!-- from default setting -->
  <a access=FULL>
      <b> 
          <c  access=READ/>
      </b>
      <x  access=READ/>
  <z>  
       <d access=FULL>
            <e access=NONE/>
       </d>
  </z>
</root>
```
Then at runtime these example authorizations would be as follows:
```
authorizeFileScope("a::b::c::more::evenmore") ==> READ
authorizeFileScope("a::b::m") ==> FULL
authorizeFileScope("z::o1::o2") ==> NONE (DEFAULT)
authorizeFileScope("z::d::e::another") ==> NONE (From Tree)
```
- Walk the tree based on the nodes of the scope path being authorized.
- Saving the last found access attribute.
- When you have a scope node that isn't in the tree at the current level, you return the last found access flag.

wangkx replied:
Will store in PTree.

afishbeck reviewed (changes requested):
@wangkx some comments.  Perhaps you could make these changes and then after that we could get others to take a look and see if this meets their needs.  We can also get suggestions on the name.

wangkx committed:
HPCC-27122 Revise based on review

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -330,6 +331,21 @@ EspHttpBinding::EspHttpBinding(IPropertyTree* tree, const char *bindname, const
     setABoolHash(proc_cfg->queryProp("@urlAlias"), serverAlias);
 }
 
+void EspHttpBinding::addServiceType(IPropertyTree* tree, const char* procName, IPropertyTree* bndCfg)
```
Will do

wangkx commented on esp/platform/application_config.cpp:
```diff
@@ -166,11 +167,19 @@ bool addSecurity(IPropertyTree *legacyEsp, IPropertyTree *appEsp, StringBuffer &
     IPropertyTree *appAuth = nullptr;
     if (isEmptyString(auth) || streq(auth, "ldap") || streq(auth, "azure_ldap"))
         appAuth = app->queryPropTree("ldap");
+    else if (streq(auth, "testauth"))
```
Will do

wangkx commented on helm/hpcc/values.yaml:
```diff
@@ -481,7 +485,7 @@ esp:
   ## Pre-configured esp applications include eclwatch, eclservices, and eclqueries
   application: eclwatch
-  auth: none
+  auth: testauth
```
Will change

wangkx commented on esp/applications/eclwatch/testauth_authorization_map.yaml:
```diff
@@ -0,0 +1,44 @@
+      # Define which authorization resources to be tested.
+      # As default, the authorization resources defined in
+      # ldap_authorization_map.yaml are tested.
+      resources: ldap
```
Will change

wangkx committed:
HPCC-27122 Set AuthDomains based on settings and remove queryTestAuthResources()

The queryTestAuthResources() should be removed in the previous commit.

Will create another PR to configure the parameters inside an
AuthDomain.

No AuthDomains is configured for the existing LdapSecurity. As
a default, the AuthTypeMixed type is used. Should it be configurable
(in another PR)?

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@afishbeck please review the last 2 commits and let me know if I miss anything. The second commit supposes to fix the TODO line at the line 141 of the application_config.cpp.

afishbeck commented:
@wangkx I think the title should be changed... from "Add Test Auth Service to ECLWatch" to "Create Test Auth secmgr for EclWatch"

wangkx commented:
@afishbeck the title is changed.

afishbeck commented on esp/platform/application_config.cpp:
```diff
@@ -89,6 +89,23 @@ static void copyDirectories(IPropertyTree *target, IPropertyTree *src)
 
+void addAuthDomains(const char *authtype, IPropertyTree *legacyEsp)
+{
+    if (isEmptyString(authtype)) //backward compatible
```
If only AuthDomain/@authType changes, you can simplify this:

```
  if (isEmptyString(authtype))
      authtype = "AuthPerRequestOnly";

 VStringBuffer AuthDomains(
            "<AuthDomains>"
                "<AuthDomain authType='%s' clientSessionTimeoutMinutes='120' domainName='default'"
                    " invalidURLsAfterAuth='/esp/login' loginLogoURL='/esp/files/eclwatch/img/Loginlogo.png'"
                    " logonURL='/esp/files/Login.html' logoutURL='' serverSessionTimeoutMinutes='240'"
                    " unrestrictedResources='/favicon.ico,/esp/files/*,/esp/xslt/*'/>"
            "</AuthDomains>", authtype);
        legacyEsp->addPropTree("AuthDomains", createPTreeFromXMLString(AuthDomains));
```

wangkx replied:
Fixed

afishbeck commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,572 @@
+    match=>continue if read permissions (if no read, implies can't "see" child scopes)
+    etc. Until full scope path checked, or no read permissions hit on ancestor scope.
+    */
+    SecAccessFlags queryFileScopeAccess(const char* scope)
```
I was picturing something a little more like the following (untested pseudo code):

```
SecAccessFlags getFileScopeAccess(IPropertyTree *accessTree, const StringArray &nodes, unsigned pos, SecAccessFlags access, bool first)
{
    if (!nodes.isItem(pos))
        return access;
    const char *node = nodes->item(pos);
    if (isEmptyString(node))
        return access;
    IPropertyTree *element = accessTree->queryPropTree(node);
    if (!element)
        return access;
    if (element->hasProp("@access"))
    {
        int newAccess = element->queryPropInt("@access);
        if (first || newAccess < access) //least access wins
        {
            access = newAccess;
            first = false;
        }
    }
    return getFileScopeAccess(element, nodes, ++pos, access, first);
}

SecAccessFlags queryFileScopeAccess(const char* scope)
{
    StringArray nodes;
    if (!convertFileScopeToNodes(scope, false, nodes))  //==>  a::b::c::d becomes nodes[a, b, c, d]
        return SecAccess_Unavailable;
    return getFileScopeAccess(fileScopeAccesses, nodes, 0, defaultFileScopeAccess, true);
}
```

Basically if no @access is found it should use the default, but as it hits each node that has an '@access' it will return the lowest found.  (I.e. if a parent is READ only you can't have a child with FULL access.)

wangkx replied:
Fixed

wangkx committed:
HPCC-27122 Revise based on review

1. Simplify the code for AuthDomains settings.
2. Improve the code for File Scope authorizations.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
Will create a new PR for review
